tests/basis_test.py

```diff
@@ -0,0 +1,161 @@
+    test_class = ChebyshevFirstKind(0)
+    for i in range(0,16):
+        assert test_class(i) == 1
+
+def test_basis_degree_1():
+    """Chebyshev polynomial degree 1 should return input"""
+    test_class = ChebyshevFirstKind(1)
+    for i in range(0,16):
+        assert test_class(i) == i
+
+def test_basis_input_1():
+    """Chebyshev polynomial input 1 returns 1 for any degree n"""
+    for i in range(0,16):
+        test_class = ChebyshevFirstKind(i)
+        assert test_class(1) == 1
+
+@pytest.mark.parametrize("x, n, expected",[(2,6,1351),(-3,9,-3880899),
+    (13,4,227137),(8,3,2024),(4,9,58106404),(14,5,8550374),(0,12,1),
+    (7,5,262087),(-2,4,97),(3,9,3880899),(9,4,51841),(2,11,978122),
+    (4,7,937444),(-3,2,17),(6,4,10081),(2,8,18817)])
+def test_basis_random_points(x,n,expected):
+    """Test chebyshev polynomial at some degree at some input"""
+    test_class = ChebyshevFirstKind(n)
+    assert test_class(x) == expected
+
+def test_is_basis_abstract():
+    """Check BasisFunction is an abstract class"""
+    with pytest.raises(TypeError):
+        test_class = BasisFunction()
+
+def test_set_initialize_empty():
+    """Check BasisFunctionSet initializes with empty set"""
+    basis_functions = []
+    points = []
+    test_class = BasisFunctionSet(points,basis_functions)
+    assert test_class.points == []
+    assert test_class.basis_functions == []
+
+def test_set_initialize_0():
+    """Check NestedBasisFunctionSet correctly initializes"""
+    basis_functions = [ChebyshevFirstKind(0)]
+    points = [0]
+    test_class = BasisFunctionSet(points,basis_functions)
+    assert test_class.points == [0]
+    assert test_class.basis_functions[0].n == 0
+
+def test_set_invalid_input():
+    """Check BasisFunctionSet gives error for invalid inputs"""
+    basis_functions = [ChebyshevFirstKind(0)]
+    points = [0,1,2]
+    with pytest.raises(IndexError):
+        test_class = BasisFunctionSet(points,basis_functions)
+
+def test_set_nested_initialize_empty():
+    """Check NestedBasisFunctionSet initializes with empty set"""
+    levels = []
+    basis_functions = []
+    points = []
+    test_class = NestedBasisFunctionSet(points,basis_functions,levels)
+    assert test_class.points == []
+    assert test_class.levels == []
+    assert test_class.basis_functions == []
+
+def test_set_nested_initialize_0():
+    """Check NestedBasisFunctionSet correctly initializes"""
+    levels = [[0]]
+    points = [0]
+    test_class = NestedBasisFunctionSet(points,[ChebyshevFirstKind(0)],levels)
+    assert test_class.points == [0]
+    assert test_class.levels == [[0]]
+
+def test_set_nested_change_levels():
+    """Check NestedBasisFunctionSet updates points when level changes"""
+    levels = [[0]]
+    points = [0]
+    test_class = NestedBasisFunctionSet(points,[ChebyshevFirstKind(0)],levels)
+    test_class.levels = [[0],[1,2],[3,4]]
+    assert test_class.levels == [[0],[1,2],[3,4]]
```

Comment:
This test would produce an unusable `NestedBasisFunctionSet`, which actually points out something you should go back and look at. The indexes that are used in `levels` must all be smaller than the length of the points/basis functions. Otherwise, you potentially get an index error. You should fix this in the main code, then write some tests that both test for the error and also will work correctly.